File: go/uri/uri.go
// SPDX-License-Identifier: MIT

// Package uri provides the Aether URI scheme — the canonical addressing format
// for resources on the Aether mesh.
//
// Grammar (ABNF, RFC 5234)
//
//	aether-uri   = "aether://" authority [ "/" path ] [ "?" query ] [ "#" fragment ]
//	authority    = aether-tag / uhid
//	aether-tag   = 5(crockford) [ "-" ] 5(crockford)         ; case-insensitive
//	uhid         = 64(HEXDIG)                                ; SHA-256 hex of public key
//	path         = path-segment *( "/" path-segment )
//	path-segment = 1*( unreserved / pct-encoded / sub-delims / ":" / "@" )
//	query        = query-param *( "&" query-param )
//	query-param  = key [ "=" value ]
//	key          = 1*( unreserved / pct-encoded )
//	value        = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
//	fragment     = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
//
// Components
//
//   - Scheme is always "aether". Case-insensitive on parse, lowercase on emit.
//   - Authority identifies the destination — an AetherTag (10 Crockford
//     base-32 chars, dash optional) or a UHID (64 hex chars). Case-insensitive.
//   - Path is opaque to the protocol — it names a handler within the
//     destination (e.g. /content/<hash>, /profile, /inbox). Case-sensitive.
//   - Query carries handler arguments. Keys are case-insensitive, values are
//     case-sensitive.
//   - Fragment is a client-side hint and is never transmitted over the wire.
//
// Why no userinfo or port
//
// The authority IS the user — there is no separate userinfo. Ports have no
// meaning in mesh routing because the transport layer selects the carrier
// (BLE / Wi-Fi Direct / NearLink / HTTP relay); a URI never picks one.
package uri

import (
	"errors"
	"fmt"
	"strings"

	"github.com/bhengubv/aether-protocol/go/identity"
)

// Scheme is the fixed scheme name — "aether".
const Scheme = "aether"

// schemePrefix is the literal scheme + "://" prefix.
const schemePrefix = "aether://"

// ParseError is returned by Parse when the input is not a valid aether URI.
type ParseError struct {
	// Input is the original string that failed to parse.
	Input string
	// Msg is a human-readable description of what went wrong.
	Msg string
}

// Error implements the error interface.
func (e *ParseError) Error() string {
	if e.Input == "" {
		return "aether-uri: " + e.Msg
	}
	return "aether-uri: " + e.Msg + " (input=" + e.Input + ")"
}

// URI is a parsed aether:// URI. The zero value is not valid; use Parse or a
// Builder to construct one.
type URI struct {
	// Authority is the destination — an AetherTag in canonical "XXXXX-XXXXX"
	// form or an upper-case 64-char hex UHID.
	Authority string

	// Path is the handler path, without the leading slash. Empty string means
	// "root" (no path).
	Path string

	// Query holds decoded query parameters. Keys are stored lower-case for
	// case-insensitive lookup; values are case-sensitive. A flag-style
	// parameter ("?flag") is represented as an entry with an empty value.
	Query map[string]string

	// Fragment is the fragment, with the leading "#" stripped. Empty if none.
	Fragment string

	// queryOrder preserves the insertion order of query keys so the canonical
	// encoder emits them in the same order as the source. Unexported because
	// it is an implementation detail of the canonical-form encoder; it does
	// NOT participate in Equal.
	queryOrder []string
}

// IsValid reports whether u was produced by a successful parse (i.e. has a
// non-empty authority).
func (u URI) IsValid() bool {
	return u.Authority != ""
}

// HandlerName returns the first path segment — the handler name — or the
// empty string for the root path.
func (u URI) HandlerName() string {
	if u.Path == "" {
		return ""
	}
	if slash := strings.IndexByte(u.Path, '/'); slash >= 0 {
		return u.Path[:slash]
	}
	return u.Path
}

// PathSegments returns the path split into segments after percent-decoding
// each one. Returns an empty slice for the root path.
func (u URI) PathSegments() []string {
	if u.Path == "" {
		return []string{}
	}
	return strings.Split(u.Path, "/")
}

// Equal reports whether two URIs have the same authority, path, fragment, and
// query map (key case-insensitive, order-insensitive).
func (u URI) Equal(other URI) bool {
	if u.Authority != other.Authority {
		return false
	}
	if u.Path != other.Path {
		return false
	}
	if u.Fragment != other.Fragment {
		return false
	}
	if len(u.Query) != len(other.Query) {
		return false
	}
	// Keys are already stored lower-case so direct comparison works.
	for k, v := range u.Query {
		ov, ok := other.Query[k]
		if !ok || ov != v {
			return false
		}
	}
	return true
}

// String returns the canonical string form of the URI. Two URIs that compare
// Equal produce the same canonical string except that query parameter order
// is not part of equality but IS part of the rendered form (insertion order
// is preserved by the Builder; Parse uses the order they appeared in the
// source). The encoder applies RFC 3986 percent-encoding using the per-kind
// allow-lists in IsAllowedUnencoded.
func (u URI) String() string {
	if !u.IsValid() {
		return ""
	}
	var sb strings.Builder
	sb.Grow(64)
	sb.WriteString(schemePrefix)
	sb.WriteString(u.Authority)
	if u.Path != "" {
		sb.WriteByte('/')
		encodePath(&sb, u.Path)
	}
	if len(u.Query) > 0 {
		sb.WriteByte('?')
		first := true
		for _, k := range u.queryKeyOrder() {
			if !first {
				sb.WriteByte('&')
			}
			first = false
			encodeComponent(&sb, k, encodeQueryKey)
			v := u.Query[k]
			if v != "" {
				sb.WriteByte('=')
				encodeComponent(&sb, v, encodeQueryValue)
			}
		}
	}
	if u.Fragment != "" {
		sb.WriteByte('#')
		encodeComponent(&sb, u.Fragment, encodeFragment)
	}
	return sb.String()
}

// queryKeyOrder returns the query keys in insertion order if a slice of
// preserved keys is attached, otherwise sorted alphabetically for
// determinism. The Parse / Builder code uses preservedKeys via a per-URI
// hidden slice; here we just produce a stable order by iterating the map.
//
// For the corpus the rendered order matches the input order because the C#
// reference iterates its Dictionary insertion order. Go's map iteration is
// randomised, so we keep insertion order explicitly via the orderedQueryKeys
// field on the URI (populated by Parse / Builder).
func (u URI) queryKeyOrder() []string {
	if u.queryOrder != nil {
		// Filter to only keys still present (RemoveQuery in builder may have
		// trimmed entries from Query but left them in queryOrder).
		out := make([]string, 0, len(u.Query))
		seen := make(map[string]struct{}, len(u.Query))
		for _, k := range u.queryOrder {
			if _, ok := u.Query[k]; ok {
				if _, dup := seen[k]; dup {
					continue
				}
				seen[k] = struct{}{}
				out = append(out, k)
			}
		}
		// Append any keys that were never registered in the order list (e.g.
		// map-only construction).
		if len(out) != len(u.Query) {
			for k := range u.Query {
				if _, ok := seen[k]; !ok {
					seen[k] = struct{}{}
					out = append(out, k)
				}
			}
		}
		return out
	}
	keys := make([]string, 0, len(u.Query))
	for k := range u.Query {
		keys = append(keys, k)
	}
	// Sort lexicographically — stable across runs.
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0 && keys[j-1] > keys[j]; j-- {
			keys[j-1], keys[j] = keys[j], keys[j-1]
		}
	}
	return keys
}

// Parse parses an aether:// URI. Returns a *ParseError on any syntactic
// violation.
func Parse(input string) (URI, error) {
	if input == "" {
		return URI{}, &ParseError{Input: input, Msg: "input is empty"}
	}

	// Scheme is case-insensitive per RFC 3986.
	if len(input) < len(schemePrefix) ||
		!strings.EqualFold(input[:len(schemePrefix)], schemePrefix) {
		return URI{}, &ParseError{Input: input, Msg: "scheme must be 'aether://'"}
	}

	rest := input[len(schemePrefix):]

	// Split on fragment first (only one '#' is allowed).
	var fragment string
	if fragSplit := strings.IndexByte(rest, '#'); fragSplit >= 0 {
		fragRaw := rest[fragSplit+1:]
		decoded, err := percentDecode(fragRaw)
		if err != nil {
			return URI{}, &ParseError{Input: input, Msg: "malformed fragment: " + err.Error()}
		}
		fragment = decoded
		rest = rest[:fragSplit]
	}

	// Then query.
	query := map[string]string{}
	var queryOrder []string
	if querySplit := strings.IndexByte(rest, '?'); querySplit >= 0 {
		queryRaw := rest[querySplit+1:]
		rest = rest[:querySplit]
		for _, pair := range strings.Split(queryRaw, "&") {
			if pair == "" {
				continue
			}
			var key, value string
			if eq := strings.IndexByte(pair, '='); eq >= 0 {
				key = pair[:eq]
				value = pair[eq+1:]
			} else {
				key = pair
			}
			decodedKey, err := percentDecode(key)
			if err != nil {
				return URI{}, &ParseError{Input: input, Msg: "malformed query key: " + err.Error()}
			}
			if decodedKey == "" {
				return URI{}, &ParseError{Input: input, Msg: "empty query parameter key"}
			}
			decodedValue, err := percentDecode(value)
			if err != nil {
				return URI{}, &ParseError{Input: input, Msg: "malformed query value: " + err.Error()}
			}
			// Match C# behaviour: keys are stored lower-case for case-insensitive
			// lookup. Later occurrences of the same key overwrite earlier ones.
			lower := strings.ToLower(decodedKey)
			if _, exists := query[lower]; !exists {
				queryOrder = append(queryOrder, lower)
			}
			query[lower] = decodedValue
		}
	}

	// Then path.
	var authorityRaw, pathRaw string
	if pathSplit := strings.IndexByte(rest, '/'); pathSplit >= 0 {
		authorityRaw = rest[:pathSplit]
		pathRaw = rest[pathSplit+1:]
	} else {
		authorityRaw = rest
	}

	if authorityRaw == "" {
		return URI{}, &ParseError{Input: input, Msg: "authority is missing"}
	}

	authority, err := canonicaliseAuthority(authorityRaw)
	if err != nil {
		return URI{}, &ParseError{Input: input, Msg: err.Error()}
	}

	if err := validatePath(pathRaw); err != nil {
		return URI{}, &ParseError{Input: input, Msg: err.Error()}
	}

	decodedPath, err := percentDecodePath(pathRaw)
	if err != nil {
		return URI{}, &ParseError{Input: input, Msg: "malformed path: " + err.Error()}
	}

	return URI{
		Authority:  authority,
		Path:       decodedPath,
		Query:      query,
		Fragment:   fragment,
		queryOrder: queryOrder,
	}, nil
}

// canonicaliseAuthority validates and canonicalises an authority. It accepts
// either a 64-char hex UHID (rendered upper-case) or an AetherTag (rendered
// in canonical "XXXXX-XXXXX" form via the identity package).
func canonicaliseAuthority(raw string) (string, error) {
	if len(raw) == 64 && isHex(raw) {
		return strings.ToUpper(raw), nil
	}
	if tag, ok := identity.TryParse(raw); ok {
		return tag.Value, nil
	}
	return "", fmt.Errorf("authority %q is neither a valid AetherTag nor a 64-char hex UHID", raw)
}

// isHex reports whether s consists entirely of hex digits.
func isHex(s string) bool {
	for i := 0; i < len(s); i++ {
		if !isHexChar(s[i]) {
			return false
		}
	}
	return true
}

// validatePath walks the path segments and rejects empty segments, illegal
// characters, and malformed percent-encoding.
func validatePath(path string) error {
	if path == "" {
		return nil
	}
	for _, segment := range strings.Split(path, "/") {
		if segment == "" {
			return errors.New("empty path segment (consecutive slashes)")
		}
		for i := 0; i < len(segment); i++ {
			c := segment[i]
			if isUnreserved(c) || isSubDelim(c) || c == ':' || c == '@' {
				continue
			}
			if c == '%' {
				if i+2 >= len(segment) || !isHexChar(segment[i+1]) || !isHexChar(segment[i+2]) {
					return fmt.Errorf("malformed percent-encoding at position %d of segment %q", i, segment)
				}
				i += 2
				continue
			}
			return fmt.Errorf("illegal character %q in path segment %q", c, segment)
		}
	}
	return nil
}

// ── character classes (RFC 3986) ─────────────────────────────────────────────

func isUnreserved(c byte) bool {
	return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') ||
		c == '-' || c == '.' || c == '_' || c == '~'
}

func isSubDelim(c byte) bool {
	switch c {
	case '!', '$', '&', '\'', '(', ')', '*', '+', ',', ';', '=':
		return true
	}
	return false
}

func isHexChar(c byte) bool {
	return (c >= '0' && c <= '9') || (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f')
}

func hexValue(c byte) int {
	switch {
	case c <= '9':
		return int(c - '0')
	case c <= 'F':
		return int(c-'A') + 10
	default:
		return int(c-'a') + 10
	}
}

// ── percent-encoding ─────────────────────────────────────────────────────────

// encodeKind selects the per-component allow-list used by encodeComponent.
type encodeKind int

const (
	encodePathSegment encodeKind = iota
	encodeQueryKey
	encodeQueryValue
	encodeFragment
)

// encodePath writes a slash-separated path, encoding each segment under the
// path-segment allow-list.
func encodePath(sb *strings.Builder, path string) {
	first := true
	for _, segment := range strings.Split(path, "/") {
		if !first {
			sb.WriteByte('/')
		}
		first = false
		encodeComponent(sb, segment, encodePathSegment)
	}
}

// encodeComponent percent-encodes value under the given kind's allow-list.
// Non-allowed code points are UTF-8 encoded and each byte is emitted as %XX.
func encodeComponent(sb *strings.Builder, value string, kind encodeKind) {
	// Iterate runes so multi-byte UTF-8 characters are encoded byte-by-byte.
	for _, r := range value {
		if r < 0x80 {
			c := byte(r)
			if isAllowedUnencoded(c, kind) {
				sb.WriteByte(c)
				continue
			}
			writeHex(sb, c)
			continue
		}
		// Multi-byte UTF-8 — always percent-encode every byte.
		buf := []byte(string(r))
		for _, b := range buf {
			writeHex(sb, b)
		}
	}
}

// writeHex writes one %XX escape using upper-case hex digits.
func writeHex(sb *strings.Builder, b byte) {
	const hex = "0123456789ABCDEF"
	sb.WriteByte('%')
	sb.WriteByte(hex[b>>4])
	sb.WriteByte(hex[b&0x0F])
}

// isAllowedUnencoded reports whether c may be emitted literally in the given
// component. The allow-lists match the C# reference's EncodeKind table.
func isAllowedUnencoded(c byte, kind encodeKind) bool {
	if isUnreserved(c) {
		return true
	}
	switch kind {
	case encodePathSegment:
		// pchar = unreserved / pct-encoded / sub-delims / ":" / "@"
		return isSubDelim(c) || c == ':' || c == '@'
	case encodeQueryKey:
		// Always encode '&' and '=' in keys; allow ':' '@' and the sub-delims
		// that do not collide with query syntax.
		switch c {
		case ':', '@', '!', '$', '\'', '(', ')', '*', '+', ',', ';':
			return true
		}
		return false
	case encodeQueryValue:
		// Allow sub-delims except '&'; '=' is fine inside a value.
		switch c {
		case ':', '@', '/', '?', '!', '$', '\'', '(', ')', '*', '+', ',', ';', '=':
			return true
		}
		return false
	case encodeFragment:
		// fragment = *( pchar / "/" / "?" )  ; pchar incl. ':' '@' sub-delims
		return isSubDelim(c) || c == ':' || c == '@' || c == '/' || c == '?'
	}
	return false
}

// percentDecode decodes %XX escapes in s, treating bytes that fall outside
// any escape as their UTF-8 representation. Returns an error if a stray '%'
// is not followed by two hex digits.
func percentDecode(input string) (string, error) {
	if strings.IndexByte(input, '%') < 0 {
		return input, nil
	}
	buf := make([]byte, 0, len(input))
	for i := 0; i < len(input); i++ {
		c := input[i]
		if c == '%' {
			if i+2 >= len(input) || !isHexChar(input[i+1]) || !isHexChar(input[i+2]) {
				return "", fmt.Errorf("malformed percent-encoding at position %d", i)
			}
			buf = append(buf, byte((hexValue(input[i+1])<<4)|hexValue(input[i+2])))
			i += 2
			continue
		}
		// A non-escaped byte may be ASCII or part of a UTF-8 sequence;
		// emit it verbatim — Go strings are already UTF-8.
		buf = append(buf, c)
	}
	return string(buf), nil
}

// percentDecodePath decodes each '/'-separated segment independently so any
// %2F bytes inside a segment are preserved as literal '/' in the decoded
// segment, while the structural slashes between segments are not.
func percentDecodePath(path string) (string, error) {
	if path == "" {
		return path, nil
	}
	segs := strings.Split(path, "/")
	for i, s := range segs {
		decoded, err := percentDecode(s)
		if err != nil {
			return "", err
		}
		segs[i] = decoded
	}
	return strings.Join(segs, "/"), nil
}

File: rust/src/uri/uri.rs
// SPDX-License-Identifier: MIT

//! `AetherUri` — the canonical addressing format for resources on the Aether mesh.
//!
//! # Grammar (ABNF, RFC 5234)
//!
//! ```text
//! aether-uri   = "aether://" authority [ "/" path ] [ "?" query ] [ "#" fragment ]
//! authority    = aether-tag / uhid
//! aether-tag   = 5(crockford) [ "-" ] 5(crockford)         ; case-insensitive
//! uhid         = 64(HEXDIG)                                ; SHA-256 hex of public key
//! path         = path-segment *( "/" path-segment )
//! path-segment = 1*( unreserved / pct-encoded / sub-delims / ":" / "@" )
//! query        = query-param *( "&" query-param )
//! query-param  = key [ "=" value ]
//! key          = 1*( unreserved / pct-encoded )
//! value        = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
//! fragment     = *( unreserved / pct-encoded / sub-delims / ":" / "@" / "/" / "?" )
//! ```
//!
//! See `docs/aether-uri-scheme.md` for the full design specification.
//!
//! # Component model
//!
//! - **Authority** — the destination, either an [`AetherNetTag`](crate::identity::AetherNetTag)
//!   (10 Crockford chars, dash optional, case-insensitive) or a 64-char hex UHID.
//!   Canonicalised to upper case on parse.
//! - **Path** — opaque to the protocol; first segment names a handler.
//!   Stored decoded; case-sensitive.
//! - **Query** — handler arguments. Keys are stored lower-case for case-insensitive
//!   lookup; values preserve case.
//! - **Fragment** — a client-side hint, stored decoded.
//!
//! # Departure from C# reference
//!
//! The query is backed by a [`BTreeMap`], so canonical emission is in
//! lexicographic key order rather than the C# `Dictionary` insertion order.
//! Two URIs with the same set of `(key, value)` pairs are byte-equal on
//! [`Display`] regardless of input order. Equality is order-insensitive in
//! both impls.

use std::collections::BTreeMap;
use std::fmt;

use crate::identity::AetherNetTag;

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/// The fixed scheme name — `aether`.
pub const SCHEME: &str = "aether";

/// The full scheme prefix — `aether://`.
const SCHEME_PREFIX: &str = "aether://";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// A parsed `aether://` URI.
///
/// All components are stored decoded and canonicalised; calling [`Display`]
/// re-encodes to RFC-3986-compliant percent-encoded form. The canonical form
/// is stable: `parse(s).to_string()` and `parse(parse(s).to_string()).to_string()`
/// produce byte-equal output.
#[derive(Clone, Debug)]
pub struct AetherUri {
    authority: String,
    path: String,
    query: BTreeMap<String, String>,
    fragment: String,
}

/// Errors that can occur when parsing, building, or dispatching an `AetherUri`.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum AetherUriError {
    /// Input was empty or `None`.
    #[error("input is null or empty")]
    Empty,

    /// Input does not begin with `aether://` (case-insensitive).
    #[error("scheme must be 'aether://'")]
    InvalidScheme,

    /// The authority slot was empty (e.g. `aether:///profile`).
    #[error("authority is missing")]
    MissingAuthority,

    /// The authority was neither a valid AetherTag nor a 64-char hex UHID.
    #[error("authority '{0}' is neither a valid AetherTag nor a 64-char hex UHID")]
    InvalidAuthority(String),

    /// The path contained an empty segment (consecutive slashes).
    #[error("empty path segment (consecutive slashes)")]
    EmptyPathSegment,

    /// The path contained a character outside the allowed set.
    #[error("illegal character '{ch}' in path segment '{segment}'")]
    IllegalPathChar { ch: char, segment: String },

    /// A `%` was not followed by two hex digits.
    #[error("malformed percent-encoding at position {pos} of segment '{segment}'")]
    MalformedPercentEncoding { pos: usize, segment: String },

    /// A query parameter had an empty key (e.g. `?=value`).
    #[error("empty query parameter key")]
    EmptyQueryKey,

    /// The builder was finalised without setting an authority.
    #[error("authority is required")]
    AuthorityRequired,

    /// A handler callback returned an error or the dispatcher encountered an
    /// internal failure. The wrapped string carries the underlying message.
    #[error("dispatch failure: {0}")]
    Dispatch(String),

    /// The descriptor passed to `Router::register` was not present in the
    /// manifest, or had an out-of-range index.
    #[error("handler index {0} is out of range for the manifest")]
    HandlerIndexOutOfRange(usize),
}

// ---------------------------------------------------------------------------
// Impl: AetherUri
// ---------------------------------------------------------------------------

impl AetherUri {
    // -----------------------------------------------------------------------
    // Constructors
    // -----------------------------------------------------------------------

    /// Parses an `aether://` URI string. Returns an [`AetherUriError`] on any
    /// syntactic violation.
    ///
    /// `parse` and `try_parse` are equivalent — both return `Result`. The two
    /// names exist for symmetry with the C# `Parse` / `TryParse` API surface.
    pub fn parse(input: &str) -> Result<Self, AetherUriError> {
        parse_internal(input)
    }

    /// Alias for [`parse`](Self::parse). Returns the same `Result`.
    pub fn try_parse(input: &str) -> Result<Self, AetherUriError> {
        parse_internal(input)
    }

    // -----------------------------------------------------------------------
    // Accessors
    // -----------------------------------------------------------------------

    /// The destination authority, canonicalised (upper-case AetherTag or
    /// upper-case 64-char hex UHID).
    pub fn authority(&self) -> &str {
        &self.authority
    }

    /// The path component without the leading slash. Empty string for root.
    pub fn path(&self) -> &str {
        &self.path
    }

    /// The fragment with the leading `#` stripped. Empty if none.
    pub fn fragment(&self) -> &str {
        &self.fragment
    }

    /// The decoded query parameters. Keys are lower-case.
    pub fn query(&self) -> &BTreeMap<String, String> {
        &self.query
    }

    /// The first path segment — i.e. the handler name. Empty for root.
    pub fn handler_name(&self) -> &str {
        if self.path.is_empty() {
            ""
        } else {
            match self.path.find('/') {
                Some(i) => &self.path[..i],
                None => &self.path,
            }
        }
    }

    /// The path split into segments (already decoded). Empty for root.
    pub fn path_segments(&self) -> Vec<&str> {
        if self.path.is_empty() {
            Vec::new()
        } else {
            self.path.split('/').collect()
        }
    }
}

// ---------------------------------------------------------------------------
// Impl: Display (canonical encoder)
// ---------------------------------------------------------------------------

impl fmt::Display for AetherUri {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(SCHEME_PREFIX)?;
        f.write_str(&self.authority)?;
        if !self.path.is_empty() {
            f.write_str("/")?;
            encode_path(f, &self.path)?;
        }
        if !self.query.is_empty() {
            f.write_str("?")?;
            let mut first = true;
            // BTreeMap iterates in lexicographic key order — stable, canonical.
            for (k, v) in &self.query {
                if !first {
                    f.write_str("&")?;
                }
                first = false;
                encode_component(f, k, EncodeKind::QueryKey)?;
                if !v.is_empty() {
                    f.write_str("=")?;
                    encode_component(f, v, EncodeKind::QueryValue)?;
                }
            }
        }
        if !self.fragment.is_empty() {
            f.write_str("#")?;
            encode_component(f, &self.fragment, EncodeKind::Fragment)?;
        }
        Ok(())
    }
}

// ---------------------------------------------------------------------------
// Impl: PartialEq / Eq (query-order-insensitive by virtue of BTreeMap)
// ---------------------------------------------------------------------------

impl PartialEq for AetherUri {
    fn eq(&self, other: &Self) -> bool {
        self.authority == other.authority
            && self.path == other.path
            && self.fragment == other.fragment
            && self.query == other.query
    }
}

impl Eq for AetherUri {}

// ---------------------------------------------------------------------------
// Parser
// ---------------------------------------------------------------------------

fn parse_internal(input: &str) -> Result<AetherUri, AetherUriError> {
    if input.is_empty() {
        return Err(AetherUriError::Empty);
    }

    // Scheme prefix is case-insensitive per RFC 3986.
    if input.len() < SCHEME_PREFIX.len()
        || !input[..SCHEME_PREFIX.len()].eq_ignore_ascii_case(SCHEME_PREFIX)
    {
        return Err(AetherUriError::InvalidScheme);
    }

    let mut rest = &input[SCHEME_PREFIX.len()..];

    // Split on the first `#` (fragment) — only one is allowed.
    let fragment = match rest.find('#') {
        Some(i) => {
            let f = percent_decode(&rest[i + 1..]);
            rest = &rest[..i];
            f
        }
        None => String::new(),
    };

    // Then split on the first `?` (query).
    let mut query: BTreeMap<String, String> = BTreeMap::new();
    if let Some(i) = rest.find('?') {
        let query_raw = &rest[i + 1..];
        rest = &rest[..i];
        for pair in query_raw.split('&') {
            if pair.is_empty() {
                continue;
            }
            let (key_raw, value_raw) = match pair.find('=') {
                Some(j) => (&pair[..j], &pair[j + 1..]),
                None => (pair, ""),
            };
            let decoded_key = percent_decode(key_raw);
            if decoded_key.is_empty() {
                return Err(AetherUriError::EmptyQueryKey);
            }
            // Keys are case-insensitive — store lower-case.
            query.insert(decoded_key.to_ascii_lowercase(), percent_decode(value_raw));
        }
    }

    // Then split on the first `/` (path).
    let (authority_raw, path_raw) = match rest.find('/') {
        Some(i) => (&rest[..i], &rest[i + 1..]),
        None => (rest, ""),
    };

    if authority_raw.is_empty() {
        return Err(AetherUriError::MissingAuthority);
    }

    let authority = canonicalise_authority(authority_raw)?;
    validate_path(path_raw)?;
    let path = percent_decode_path(path_raw);

    Ok(AetherUri { authority, path, query, fragment })
}

fn canonicalise_authority(raw: &str) -> Result<String, AetherUriError> {
    // Try UHID (64 hex chars).
    if raw.len() == 64 && is_all_hex(raw) {
        return Ok(raw.to_ascii_uppercase());
    }
    // Try AetherTag (10 Crockford chars, dash optional, case-insensitive).
    if let Ok(tag) = AetherNetTag::parse(raw) {
        return Ok(tag.value().to_string()); // already in canonical XXXXX-XXXXX upper-case.
    }
    Err(AetherUriError::InvalidAuthority(raw.to_string()))
}

fn validate_path(path: &str) -> Result<(), AetherUriError> {
    if path.is_empty() {
        return Ok(());
    }
    for segment in path.split('/') {
        if segment.is_empty() {
            return Err(AetherUriError::EmptyPathSegment);
        }
        let bytes = segment.as_bytes();
        let mut i = 0;
        while i < bytes.len() {
            let b = bytes[i];
            let c = b as char;
            if is_unreserved(c) || is_sub_delim(c) || c == ':' || c == '@' {
                i += 1;
                continue;
            }
            if c == '%' {
                if i + 2 >= bytes.len() || !is_hex_char(bytes[i + 1] as char) || !is_hex_char(bytes[i + 2] as char) {
                    return Err(AetherUriError::MalformedPercentEncoding {
                        pos: i,
                        segment: segment.to_string(),
                    });
                }
                i += 3;
                continue;
            }
            return Err(AetherUriError::IllegalPathChar {
                ch: c,
                segment: segment.to_string(),
            });
        }
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Percent-encoding tables and codecs
// ---------------------------------------------------------------------------

/// Where a character is being encoded — determines the allowed-unencoded set.
#[derive(Clone, Copy)]
enum EncodeKind {
    PathSegment,
    QueryKey,
    QueryValue,
    Fragment,
}

fn is_unreserved(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '-' || c == '.' || c == '_' || c == '~'
}

fn is_sub_delim(c: char) -> bool {
    matches!(
        c,
        '!' | '$' | '&' | '\'' | '(' | ')' | '*' | '+' | ',' | ';' | '='
    )
}

fn is_hex_char(c: char) -> bool {
    c.is_ascii_hexdigit()
}

fn is_all_hex(s: &str) -> bool {
    s.bytes().all(|b| (b as char).is_ascii_hexdigit())
}

fn is_allowed_unencoded(c: char, kind: EncodeKind) -> bool {
    if is_unreserved(c) {
        return true;
    }
    match kind {
        EncodeKind::PathSegment => is_sub_delim(c) || c == ':' || c == '@',
        EncodeKind::QueryKey => matches!(
            c,
            ':' | '@' | '!' | '$' | '\'' | '(' | ')' | '*' | '+' | ',' | ';'
        ),
        EncodeKind::QueryValue => matches!(
            c,
            ':' | '@' | '/' | '?' | '!' | '$' | '\'' | '(' | ')' | '*' | '+' | ',' | ';' | '='
        ),
        EncodeKind::Fragment => is_sub_delim(c) || c == ':' || c == '@' || c == '/' || c == '?',
    }
}

fn encode_path(f: &mut fmt::Formatter<'_>, path: &str) -> fmt::Result {
    let mut first = true;
    for segment in path.split('/') {
        if !first {
            f.write_str("/")?;
        }
        first = false;
        encode_component(f, segment, EncodeKind::PathSegment)?;
    }
    Ok(())
}

fn encode_component(f: &mut fmt::Formatter<'_>, s: &str, kind: EncodeKind) -> fmt::Result {
    for c in s.chars() {
        if is_allowed_unencoded(c, kind) {
            // Write as one char.
            // SAFETY of correctness: char::encode_utf8 with a 4-byte buffer always fits.
            let mut buf = [0u8; 4];
            f.write_str(c.encode_utf8(&mut buf))?;
            continue;
        }
        // Percent-encode each UTF-8 byte as %XX (upper-case hex).
        let mut buf = [0u8; 4];
        let s = c.encode_utf8(&mut buf);
        for &b in s.as_bytes() {
            write!(f, "%{:02X}", b)?;
        }
    }
    Ok(())
}

fn percent_decode(input: &str) -> String {
    if !input.contains('%') {
        return input.to_string();
    }
    let bytes = input.as_bytes();
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        let b = bytes[i];
        if b == b'%' && i + 2 < bytes.len() && is_hex_char(bytes[i + 1] as char) && is_hex_char(bytes[i + 2] as char) {
            out.push((hex_value(bytes[i + 1]) << 4) | hex_value(bytes[i + 2]));
            i += 3;
        } else {
            // Pass through. UTF-8 already in `bytes`.
            out.push(b);
            i += 1;
        }
    }
    // Bytes are valid UTF-8 by construction (every non-decoded byte was UTF-8
    // and decoded bytes came from a valid percent-encoded UTF-8 sequence). If
    // a malformed mix produced invalid UTF-8 the input was malformed; fall
    // back to a lossy conversion rather than panicking.
    match String::from_utf8(out) {
        Ok(s) => s,
        Err(e) => String::from_utf8_lossy(&e.into_bytes()).into_owned(),
    }
}

fn percent_decode_path(path: &str) -> String {
    if path.is_empty() {
        return String::new();
    }
    // Decode each segment independently so `/` isn't lost.
    path.split('/')
        .map(percent_decode)
        .collect::<Vec<_>>()
        .join("/")
}

fn hex_value(b: u8) -> u8 {
    match b {
        b'0'..=b'9' => b - b'0',
        b'A'..=b'F' => b - b'A' + 10,
        b'a'..=b'f' => b - b'a' + 10,
        _ => 0, // is_hex_char gate prevents reaching here in practice.
    }
}
